File: engine/src/npmguard/inventory/parse_manifest.py
"""Parse package.json into structured inventory data."""

from __future__ import annotations

from typing import Any

from npmguard.inventory.models import EntryPoints, PackageMetadata

LIFECYCLE_SCRIPTS = frozenset(("preinstall", "install", "postinstall", "prepare", "prepublish"))


def extract_script_file_ref(script_value: str) -> str | None:
    """Extract the JS file path from a script like ``"node [flags] <file>"``."""
    parts = script_value.strip().split()
    if not parts or parts[0] != "node":
        return None
    for part in parts[1:]:
        if not part.startswith("-"):
            return part
    return None


def _extract_exports_entries(exports: Any) -> list[str]:
    """Recursively extract file paths from an ``exports`` field."""
    if isinstance(exports, str):
        return [exports]
    if isinstance(exports, dict):
        entries: list[str] = []
        for value in exports.values():
            entries.extend(_extract_exports_entries(value))
        return entries
    return []


def _extract_bin_entries(bin_field: Any) -> list[str]:
    if isinstance(bin_field, str):
        return [bin_field]
    if isinstance(bin_field, dict):
        return [v for v in bin_field.values() if isinstance(v, str)]
    return []


def parse_package_json(
    pkg: dict[str, Any],
) -> tuple[PackageMetadata, dict[str, str], EntryPoints, dict[str, dict[str, str]]]:
    """Parse a package.json dict into structured inventory components.

    Returns (metadata, scripts, entry_points, dependencies).
    """
    metadata = PackageMetadata(
        name=pkg.get("name"),
        version=pkg.get("version"),
        description=pkg.get("description"),
        license=pkg.get("license"),
        homepage=pkg.get("homepage"),
        repository=pkg.get("repository"),
    )

    scripts: dict[str, str] = pkg.get("scripts", {})

    install_entries: list[str] = []
    for hook in LIFECYCLE_SCRIPTS:
        if hook in scripts:
            ref = extract_script_file_ref(scripts[hook])
            if ref is not None:
                install_entries.append(ref)

    runtime_entries: list[str] = [pkg.get("main", "index.js")]
    if pkg.get("module"):
        runtime_entries.append(pkg["module"])
    runtime_entries.extend(_extract_exports_entries(pkg.get("exports")))
    runtime_entries = list(dict.fromkeys(runtime_entries))  # dedup, preserve order

    entry_points = EntryPoints(
        install=install_entries,
        runtime=runtime_entries,
        bin=_extract_bin_entries(pkg.get("bin")),
    )

    dependencies = {
        "prod": pkg.get("dependencies", {}),
        "dev": pkg.get("devDependencies", {}),
        "optional": pkg.get("optionalDependencies", {}),
        "peer": pkg.get("peerDependencies", {}),
    }

    return metadata, scripts, entry_points, dependencies

File: engine/src/inventory/parse-manifest.ts
import type { EntryPoints, PackageMetadata } from "../models.js";

export const LIFECYCLE_SCRIPTS = new Set([
  "preinstall",
  "install",
  "postinstall",
  "prepare",
  "prepublish",
]);

function asStringOrNull(value: unknown): string | null {
  return typeof value === "string" ? value : null;
}

function asStringRecord(value: unknown): Record<string, string> {
  if (!value || typeof value !== "object" || Array.isArray(value)) return {};
  const result: Record<string, string> = {};
  for (const [k, v] of Object.entries(value)) {
    if (typeof v === "string") result[k] = v;
  }
  return result;
}

export function extractScriptFileRef(scriptValue: string): string | null {
  const parts = scriptValue.trim().split(/\s+/);
  if (!parts.length || parts[0] !== "node") return null;
  for (const part of parts.slice(1)) {
    if (!part.startsWith("-")) return part;
  }
  return null;
}

function extractExportsEntries(exports: unknown): string[] {
  if (typeof exports === "string") return [exports];
  if (exports && typeof exports === "object" && !Array.isArray(exports)) {
    const entries: string[] = [];
    for (const value of Object.values(exports)) {
      entries.push(...extractExportsEntries(value));
    }
    return entries;
  }
  return [];
}

function extractBinEntries(bin: unknown): string[] {
  if (typeof bin === "string") return [bin];
  if (bin && typeof bin === "object" && !Array.isArray(bin)) {
    return Object.values(bin).filter((v): v is string => typeof v === "string");
  }
  return [];
}

export function parsePackageJson(
  pkg: Record<string, unknown>,
): {
  metadata: PackageMetadata;
  scripts: Record<string, string>;
  entryPoints: EntryPoints;
  dependencies: Record<string, Record<string, string>>;
} {
  const metadata: PackageMetadata = {
    name: asStringOrNull(pkg.name),
    version: asStringOrNull(pkg.version),
    description: asStringOrNull(pkg.description),
    license: asStringOrNull(pkg.license),
    homepage: asStringOrNull(pkg.homepage),
    repository: pkg.repository ?? null,
  };

  const scripts = asStringRecord(pkg.scripts);

  const installEntries: string[] = [];
  for (const hook of LIFECYCLE_SCRIPTS) {
    const scriptValue = scripts[hook];
    if (scriptValue) {
      const ref = extractScriptFileRef(scriptValue);
      if (ref) installEntries.push(ref);
    }
  }

  let runtimeEntries: string[] = [asStringOrNull(pkg.main) ?? "index.js"];
  const moduleEntry = asStringOrNull(pkg.module);
  if (moduleEntry) runtimeEntries.push(moduleEntry);
  runtimeEntries.push(...extractExportsEntries(pkg.exports));
  runtimeEntries = [...new Set(runtimeEntries)]; // dedup

  const entryPoints: EntryPoints = {
    install: installEntries,
    runtime: runtimeEntries,
    bin: extractBinEntries(pkg.bin),
  };

  const dependencies: Record<string, Record<string, string>> = {
    prod: asStringRecord(pkg.dependencies),
    dev: asStringRecord(pkg.devDependencies),
    optional: asStringRecord(pkg.optionalDependencies),
    peer: asStringRecord(pkg.peerDependencies),
  };

  return { metadata, scripts, entryPoints, dependencies };
}
